### PR TITLE
feat: Add more device context

### LINKED ISSUE #1235: Add more device context
Events from iOS have less device context than, for example, events from Android. We can add more.

## iOS
```JSON
"device": {
      "family": "iOS",
      "model": "iPhone13,2",
      "model_id": "D53gAP",
      "memory_size": 3883876352,
      "free_memory": 88358912,
      "usable_memory": 2271002624,
      "storage_size": 63870980096,
      "boot_time": "2021-07-19T18:15:30Z",
      "timezone": "CEST",
      "type": "device"
}
```

## Android

```JSON
"device": {
      "name": "Pixel 2 XL",
      "family": "Pixel",
      "model": "Pixel 2 XL",
      "model_id": "RP1A.201005.004.A1",
      "battery_level": 100.0,
      "orientation": "portrait",
      "manufacturer": "Google",
      "brand": "google",
      "screen_density": 3.5,
      "screen_dpi": 560,
      "online": true,
      "charging": true,
      "low_memory": false,
      "simulator": false,
      "memory_size": 3839954944,
      "free_memory": 2134429696,
      "storage_size": 55782412288,
      "free_storage": 46677245952,
      "boot_time": "2021-07-19T11:13:52.575Z",
      "timezone": "Europe/Vienna",
      "archs": [
        "arm64-v8a",
        "armeabi-v7a",
        "armeabi"
      ],
      "battery_temperature": 30.6,
      "connection_type": "wifi",
      "id": "9cd59ebb-583b-474b-8ad1-15af30ba62d1",
      "language": "en_US",
      "screen_height_pixels": 2712,
      "screen_width_pixels": 1440,
      "type": "device"
}
```

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

Adds a bunch more properties to the device context dict.

Screen Shot 2022-09-19 at 16 42 57

## :bulb: Motivation and Context

Closes #1235

## :green_heart: How did you test it?

Unit tests & ran the app on my device

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryCrashIntegration.m [hunk 1]
```diff
@@ -248,11 +250,11 @@ + (void)enrichScope:(SentryScope *)scope crashWrapper:(SentryCrashWrapper *)cras
     [deviceData setValue:systemInfo[@"freeMemory"] forKey:SentryDeviceContextFreeMemoryKey];
     [deviceData setValue:systemInfo[@"usableMemory"] forKey:@"usable_memory"];
     [deviceData setValue:systemInfo[@"memorySize"] forKey:@"memory_size"];
-    [deviceData setValue:systemInfo[@"storageSize"] forKey:@"storage_size"];
+    [deviceData setValue:systemInfo[@"totalStorageSize"] forKey:@"storage_size"];
+    [deviceData setValue:systemInfo[@"freeStorageSize"] forKey:@"free_storage"];
     [deviceData setValue:systemInfo[@"bootTime"] forKey:@"boot_time"];
-
-    NSString *locale = [[NSLocale autoupdatingCurrentLocale] objectForKey:NSLocaleIdentifier];
-    [deviceData setValue:locale forKey:LOCALE_KEY];
```
Comment (kevinrenskers) [inline Sources/Sentry/SentryCrashIntegration.m:255]: I've removed locale from the device context, as we're already sending the culture context which contains the locale as well.

Comment (philipphofmann) [reply Sources/Sentry/SentryCrashIntegration.m:257]: `h`: Please don't as locale is indexed and culture context not yet, see https://github.com/getsentry/sentry/issues/30525.

File: Sources/Sentry/include/SentryUIDeviceWrapper.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+#if SENTRY_HAS_UIKIT
+#    import <UIKit/UIKit.h>
+#endif
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SentryUIDeviceWrapper : NSObject
+
+#if TARGET_OS_IOS
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryUIDeviceWrapper.h:11]: `m`: If all that only works on iOS, I think we could wrap the whole class in this compiler directive.

File: Sources/Sentry/SentryCrashIntegration.m [hunk 3]
```diff
@@ -233,6 +233,8 @@ + (void)enrichScope:(SentryScope *)scope crashWrapper:(SentryCrashWrapper *)cras
 
 #if TARGET_OS_SIMULATOR
     [deviceData setValue:@(YES) forKey:@"simulator"];
+#else
+    [deviceData setValue:@(NO) forKey:@"simulator"];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryCrashIntegration.m:237]: Makes sense thanks 🙏 . Please add some tests for this functionality.

File: Tests/SentryTests/Integrations/SentryCrash/SentryCrashIntegrationTests.swift [hunk 4]
```diff
@@ -216,21 +216,6 @@ class SentryCrashIntegrationTests: XCTestCase {
         XCTAssertTrue(fixture.sentryCrash.closeCalled)
     }
     
-    func testUninstall_DoesNotUpdateLocale_OnLocaleDidChangeNotification() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SentryCrash/SentryCrashIntegrationTests.swift:219]: `m`: Why did you remove these tests?

File: Sources/Sentry/SentryClient.m [hunk 5]
```diff
@@ -759,12 +766,31 @@ - (void)applyCultureContextToEvent:(SentryEvent *)event
...
                   block:^(NSMutableDictionary *device) {
                       device[SentryDeviceContextFreeMemoryKey] = @(self.crashWrapper.freeMemory);
+                      device[@"free_storage"] = @(self.crashWrapper.freeStorage);
+
+#if TARGET_OS_IOS
+                      if (self.deviceWrapper.orientation != UIDeviceOrientationUnknown) {
+                          device[@"orientation"]
+                              = UIDeviceOrientationIsPortrait(self.deviceWrapper.orientation)
+                              ? @"portrait"
+                              : @"landscape";
```
Comment (philipphofmann) [inline Sources/Sentry/SentryClient.m:782]: `l`: We could add a test for this constant here. If I change this line of code, no tests fail.

**Review by philipphofmann — COMMENTED** [review]
Mostly nitpicks. Thanks, @kevinrenskers 👏

File: Sources/Sentry/SentryCrashIntegration.m (hunk 3 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryCrashIntegration.m:237]: I don't think a test would make a lot of sense? We'd always be checking for a NO value?

File: Tests/SentryTests/Integrations/SentryCrash/SentryCrashIntegrationTests.swift (hunk 4 above)
Comment (kevinrenskers) [reply Tests/SentryTests/Integrations/SentryCrash/SentryCrashIntegrationTests.swift:219]: Because I removed the locale property, which needs to be restored. So I'll restore the tests as well.

File: Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.h [hunk 6]
```diff
@@ -29,6 +29,8 @@ SENTRY_NO_INIT
 
 @property (nonatomic, assign) uint64_t internalAppMemory;
 
+@property (nonatomic, assign) uint64_t internalFreeStorage;
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.h:32]: Then we should change all of them. Right now all of these properties are in bytes, I think it's weird to rename only one to reflect that. What about using a typealias for `uint64_t`, name that `bytes`?

File: Sources/Sentry/SentryCrashIntegration.m (hunk 1 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryCrashIntegration.m:257]: Ah, my bad. Restored!

File: Sources/Sentry/SentryCrashIntegration.m [hunk 7]
```diff
@@ -250,11 +252,14 @@ + (void)enrichScope:(SentryScope *)scope crashWrapper:(SentryCrashWrapper *)cras
     [deviceData setValue:systemInfo[@"freeMemory"] forKey:SentryDeviceContextFreeMemoryKey];
     [deviceData setValue:systemInfo[@"usableMemory"] forKey:@"usable_memory"];
     [deviceData setValue:systemInfo[@"memorySize"] forKey:@"memory_size"];
-    [deviceData setValue:systemInfo[@"storageSize"] forKey:@"storage_size"];
+    [deviceData setValue:systemInfo[@"totalStorageSize"] forKey:@"storage_size"];
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryCrashIntegration.m:255]: What exactly do you want to test? There are no tests for any of the other properties either. Do you just want to validate that a mocked value is getting set?

File: Sources/Sentry/SentryClient.m (hunk 5 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryClient.m:782]: That surprises me, as there is already a test for this in SentryClientTests 🤔
I'll look into this tomorrow.

File: Sources/Sentry/SentryClient.m [hunk 8]
```diff
@@ -759,12 +766,31 @@ - (void)applyCultureContextToEvent:(SentryEvent *)event
...
+                      }
+
+                      if (self.deviceWrapper.isBatteryMonitoringEnabled) {
+                          device[@"charging"]
+                              = self.deviceWrapper.batteryState == UIDeviceBatteryStateCharging
+                              ? @(YES)
+                              : @(NO);
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryClient.m:789]: That surprises me, as there is already a test for this in SentryClientTests 🤔
I'll look into this tomorrow.

File: Sources/Sentry/SentryUIDeviceWrapper.m [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+- (void)dealloc
+{
+    if (self.cleanupDeviceOrientationNotifications) {
+        [UIDevice.currentDevice endGeneratingDeviceOrientationNotifications];
+    }
+    if (self.cleanupBatteryMonitoring) {
+        UIDevice.currentDevice.batteryMonitoringEnabled = NO;
+    }
```
Comment (brustolin) [inline Sources/Sentry/SentryUIDeviceWrapper.m:40]: h: What if the SDK inits before the user enables both orientation notification and battery monitoring. So I guess this would disabled it even if the users need it. 
I guess, in the majority of cases, this class will not be deallocated during app usage. But if you have a solution for this it would be better.

File: Sources/Sentry/SentryCrashWrapper.m [hunk 10]
```diff
@@ -76,6 +76,11 @@ - (uint64_t)freeMemory
     return sentrycrashcm_system_freememory();
 }
 
+- (uint64_t)freeStorage
```
Comment (brustolin) [inline Sources/Sentry/SentryCrashWrapper.m:79]: l: can we rename this to 'freeStorageSize'?
I was confused the first time I read this function. Same for the C function.

File: Sources/Sentry/SentryUIDeviceWrapper.m (hunk 9 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryUIDeviceWrapper.m:40]: I'm not sure I am following your thinking here. If the SentryClient inits, then we enable monitoring. The user doesn't do this. Then when the SentryClient deallocs, we clean up after ourselves.

Comment (kevinrenskers) [reply Sources/Sentry/SentryUIDeviceWrapper.m:40]: Ah I see what you mean. If we enable it, the user ALSO enables it, then we dealloc and stop the monitoring and also disable it for the user. Well yeah, not sure how we would work around that? And also not sure if this would happen in real life?

File: Sources/Sentry/SentryCrashWrapper.m (hunk 10 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryCrashWrapper.m:79]: And `freeMemory` was not confusing? I'd rather keep it consistent with existing properties.

File: Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Tests/SentryTests/SentryCrash/TestSentryCrashWrapper.h:32]: Typealias is acceptable to me, but as you said, better in a separate PR.

File: Sources/Sentry/SentryCrashIntegration.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryCrashIntegration.m:255]: Yes, exactly. If you remove this key or change its name to `storage_sze` a test should fail. It's not optimal that we don't test the other properties. Maybe your test could cover them too.

File: Sources/Sentry/SentryCrashIntegration.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryCrashIntegration.m:237]: Hmm, no test is OK as well. We would have to add another level of abstraction, but that's overkill. Let's keep it as it is.

File: Sources/Sentry/SentryClient.m (hunk 5 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m:782]: We test for `portrait` I think but not for `landscape`.

**Review by philipphofmann — COMMENTED** [review]
Just some missing tests and the comment of @brustolin. Otherwise already LGTM

File: Sources/Sentry/SentryCrashIntegration.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryCrashIntegration.m:255]: Let's add the tests in another PR, together with either new names or type aliases.

**Review by philipphofmann — APPROVED** [review]
LGTM

File: Sources/Sentry/SentryUIDeviceWrapper.m (hunk 9 above)
Comment (brustolin) [reply Sources/Sentry/SentryUIDeviceWrapper.m:40]: If this is not easily achievable, Im fine with the way it is right now. We just need to update the docs to point this out I guess.

**Comment by kevinrenskers** [comment]
17 hours ago I had successful tests, and now "Benchmarking / Build app and test runner (pull_request)" is failing with this:

```
❌  error: No profiles for 'io.sentry.sample.iOS-Swift.Clip' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'io.sentry.sample.iOS-Swift.Clip'. Automatic signing is disabled and unable to generate a profile. To enable automatic signing, pass -allowProvisioningUpdates to xcodebuild. (in target 'iOS-SwiftClip' from project 'iOS-Swift')

❌  error: No profiles for 'io.sentry.sample.iOS-Swift' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'io.sentry.sample.iOS-Swift'. Automatic signing is disabled and unable to generate a profile. To enable automatic signing, pass -allowProvisioningUpdates to xcodebuild. (in target 'iOS-Swift' from project 'iOS-Swift')
```

And it seems it's a brand new test: we used to have 49 tests, now we have 50 of them. Where was this test added? And if that was after https://github.com/getsentry/sentry-cocoa/pull/2184, why didn't the PR that added this new test fail?

**Comment by kevinrenskers** [comment]
Ok so it seems it was part of https://github.com/getsentry/sentry-cocoa/pull/2201, which did indeed have a failing test.

**Comment by kevinrenskers** [comment]
Merging even though we have this failing test (unrelated to the PR), and creating a new PR to fix the test.

**Comment by kevinrenskers** [comment]
Adds a bunch more properties to the device context dict.
